sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -159,7 +159,11 @@ case class HashAggregateExec(
 
   // The variables are used as aggregation buffers and each aggregate function has one or more
   // ExprCode to initialize its buffer slots. Only used for aggregation without keys.
-  private var bufVars: Seq[Seq[ExprCode]] = _
+  private val bufVar = new ThreadLocal[Seq[Seq[ExprCode]]] {
```

Comment:
I thought about this too,  but my concern is that we never promise the codegen to be thread-safe. We may have more places to fix, and we may break it again in the future.